pkg/controller/apicast/apicast_logic_reconciler.go

```diff
@@ -0,0 +1,431 @@
+package apicast
+
+import (
+	"context"
+	"reflect"
+
+	apicast "github.com/3scale/apicast-operator/pkg/apicast"
+	appscommon "github.com/3scale/apicast-operator/pkg/apis/apps"
+	appsv1alpha1 "github.com/3scale/apicast-operator/pkg/apis/apps/v1alpha1"
+
+	"github.com/3scale/apicast-operator/pkg/k8sutils"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	AdmPortalSecretResverAnnotation = "apicast.apps.3scale.net/admin-portal-secret-resource-version"
+)
+
+type APIcastLogicReconciler struct {
+	BaseReconciler
+	APIcastCR *appsv1alpha1.APIcast
+}
+
+func NewAPIcastLogicReconciler(b BaseReconciler, cr *appsv1alpha1.APIcast) APIcastLogicReconciler {
+	return APIcastLogicReconciler{
+		BaseReconciler: b,
+		APIcastCR:      cr,
+	}
+}
+
+func (r *APIcastLogicReconciler) namespacedNameOnCR(obj metav1.Object) types.NamespacedName {
+	return types.NamespacedName{
+		Name:      obj.GetName(),
+		Namespace: r.APIcastCR.Namespace,
+	}
+}
+
+func (r *APIcastLogicReconciler) Reconcile() (reconcile.Result, error) {
+	r.Logger().WithValues("Name", r.APIcastCR.Name, "Namespace", r.APIcastCR.Namespace)
+
+	appliedInitialization, err := r.initialize()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if appliedInitialization {
+		return reconcile.Result{}, nil
+	}
+
+	// TODO this function does a little bit of creating the desiredApicast and
+	// also validation. Also validation should be done BEFORE
+	// the initialization probably
+	desiredAPIcast, err := r.internalAPIcast()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	desiredAPIcastAdminPortalEndpointSecret, err := desiredAPIcast.AdminPortalEndpointSecret()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	err = r.reconcileAdminPortalEndpointSecret(desiredAPIcastAdminPortalEndpointSecret)
```

Comment:
reconcile admin portal secret was not done in "internalAPICast` method. This is confusing